model/riscv_vmem.sail

```diff
@@ -0,0 +1,854 @@
+    let arch = architecture(get_mstatus_SXL(mstatus));
+    match arch {
+      Some(RV64) => { let mbits : bits(4) = satp[63 .. 60];
+                      match satp64Mode_of_bits(RV64, mbits) {    // see riscv_types.sail
+                        Some(m) => m,
+                        None()  => internal_error(__FILE__, __LINE__,
+                                                  "invalid RV64 translation mode in satp")
+                      }
+                    },
+      Some(RV32) => match satp[31] {
+                      // When xlen is 64, mstatus.SXL (for S privilege) can be RV32
+                      bitzero => Sbare,
+                      bitone  => Sv32
+                    },
+      _    => internal_error(__FILE__, __LINE__, "unsupported address translation arch")
+    }
+  }
+  else
+    internal_error(__FILE__, __LINE__, "unsupported xlen")
+}
+
+// ****************************************************************
+// VA to PA translation
+
+// Result of address translation
+
+// PUBLIC
+union TR_Result('paddr : Type, 'failure : Type) = {
+  TR_Address : ('paddr, ext_ptw),
+  TR_Failure : ('failure, ext_ptw)
+}
+
+// This function can be ignored on first reading since TLBs are not
+// part of RISC-V architecture spec (see TLB_NOTE above).
+// PRIVATE: translate on TLB hit, and maintenance of PTE in TLB
+function translate_TLB_hit(sv_params : SV_Params,
+                           asid      : bits(16),
+                           ptb       : bits(64),
+                           vAddr     : bits(64),
+                           ac        : AccessType(ext_access_type),
+                           priv      : Privilege,
+                           mxr       : bool,
+                           do_sum    : bool,
+                           ext_ptw   : ext_ptw,
+                           tlb_index : nat,
+                           ent       : TLB_Entry)
+                          -> TR_Result(bits(64), PTW_Error) = {
+  let pte       = ent.pte;
+  let ext_pte   = msbs_of_PTE(sv_params, pte);
+  let pte_lsbs  = Mk_PTE_LSBs(pte[7 .. 0]);
+  let pte_check = check_PTE_permission(ac, priv, mxr, do_sum, pte_lsbs,
+                                       ext_pte,
+                                       ext_ptw);
+  let tr_result : TR_Result(bits(64), PTW_Error)
+  = match pte_check {
+      PTE_Check_Failure(ext_ptw, ext_ptw_fail) =>
+        TR_Failure(ext_get_ptw_error(ext_ptw_fail), ext_ptw),
+      PTE_Check_Success(ext_ptw) =>
+        match update_PTE_Bits(sv_params, pte, ac, ext_pte) {
+          None()     => TR_Address(ent.pAddr | (vAddr & ent.vAddrMask), ext_ptw),
+          Some(pte') =>
+            // See riscv_platform.sail
+            if not(plat_enable_dirty_update()) then
+              // pte needs dirty/accessed update but that is not enabled
+              TR_Failure(PTW_PTE_Update(), ext_ptw)
+            else {
+              // Writeback the PTE (which has new A/D bits)
+              n_ent : TLB_Entry = ent;
```

Comment:
I think this works:

```
let n_ent = {ent with pte=pte'};
```

If not, we should at least use explicit `var n_ent : ...`

Reply:
Ok. Fixed.  Was a ‘var’ only to do a field update; I was not aware that ‘with’ can be used to do it functionally. The Sail manual documents ‘with’ for vector and. bitfield updates but not for structs (in the main text, not the grammar).